ykcapi/src/lib.rs

```diff
@@ -83,33 +116,12 @@ mod c_testing {
     }
 
     #[no_mangle]
-    pub extern "C" fn __yktrace_irtrace_len(trace: *mut IRTrace) -> usize {
-        let trace = unsafe { Box::from_raw(trace) };
-        let ret = trace.len();
-        mem::forget(trace);
-        ret
-    }
-
-    /// Fetches the function name (`res_func`) and the block index (`res_bb`) at position `idx` in
-    /// `trace`.
-    #[no_mangle]
-    pub extern "C" fn __yktrace_irtrace_get(
-        trace: *mut IRTrace,
-        idx: usize,
-        res_func: *mut *const c_char,
-        res_bb: *mut usize,
-    ) {
-        let trace = unsafe { Box::from_raw(trace) };
-        let blk = trace.get(idx).unwrap();
-        unsafe {
-            *res_func = blk.func_name().as_ptr();
-            *res_bb = blk.bb();
-        }
-        mem::forget(trace);
+    pub extern "C" fn __yktrace_drop_irtrace(trace: *mut IRTrace) {
+        unsafe { Box::from_raw(trace) };
     }
 
     #[no_mangle]
-    pub extern "C" fn __yktrace_drop_irtrace(trace: *mut IRTrace) {
-        unsafe { Box::from_raw(trace) };
+    pub extern "C" fn __yktrace_irtrace_compile(trace: *mut IRTrace) {
+        unsafe {&*trace}.compile();
```

Comment:
Is this a way to get around the `Box::new()/mem::forget()` dance? If so, why don't we use it in the other places too?

Reply:
I'm not sure that's what's going on here, but it looks to me like Edd uses `{&*trace}.compile()` instead of doing
```rust
let t = Box::from_raw(trace);
t.compile();
mem::forget(t);
```
as he's done in the other functions. If that is the case, then I'm wondering why we are not consistent everywhere. Unless there's something else going on here.

Reply:
`&*` converts a `*mut` into a `&`: since it's a reference, there is nothing to drop (you can't drop references). It's much more elegant than the `from_raw`/`forget` dance -- and more flexible, because it doesn't require the pointer to have been created from a `Box`.

Reply:
Yes, hence my question why we don't use this in the other places too instead of doing `from_raw/forget`, like [here](https://github.com/ykjit/yk/pull/342/commits/b408c8a0ced85e7c0c994682f5c50f46e21bbf1f#diff-e609c9c537d75440ea3db6f14fa49100e4e511dab07bfb704cc4e8199fb1fae3R49) and [here](https://github.com/ykjit/yk/pull/342/commits/b408c8a0ced85e7c0c994682f5c50f46e21bbf1f#diff-e609c9c537d75440ea3db6f14fa49100e4e511dab07bfb704cc4e8199fb1fae3R64).

Reply:
I don't think Edd's introduced any new `from_raw`'s in this PR. I think fixing those in a separate PR would be a good thing.

Reply:
What do you mean? The two links I posted above are functions introduced in this PR, where he can also use `&*` instead of `from_raw/forget`.

Reply:
I *think* later commits have removed that code.

Reply:
Huh, I guess I'm the numpty here. I didn't look at the [overall changes](https://github.com/ykjit/yk/pull/342/files) and just the separate commits. I only saw that this code was moved around [herehttps://github.com/ykjit/yk/pull/342/commits/92039318fe5ddabc96cb787f2c3cd75e5ba9b250#diff-e609c9c537d75440ea3db6f14fa49100e4e511dab07bfb704cc4e8199fb1fae3R87) and then never touched again. So I assumed it must be still there. What I didn't realise is that this code was just moved somewhere else in the first commit, and then moved back to where it was initially in the second commit, negating those changes.

Anyway, I think it's still worth changing. Either here or in a separate PR.

Reply:
There are indeed some other places where I could use `&*` instead of all of the boxing fun. I only recently remembered this trick.

Reply:
Fixed in a commit that I'll leave separate: 241f73d